neuralpp/experiments/experimental_inference/graph_computation.py

```diff
@@ -1,88 +1,74 @@
-
-
-Expansion = namedtuple("Expansion", "node expansion_value")
-
-
-class ExpansionValueComputation(TreeComputation):
-
-    def __init__(self, partial_tree, full_tree, expansion_value_function):
-        super().__init__(partial_tree)
-        self.full_tree = full_tree
-        self.expansion_value_function = expansion_value_function
-        self.compute_result_dict(partial_tree.root)
-
-    def __compute_single_node(self, node, get_child_scores_from_cache: bool):
-        children = self.full_tree.children(node)
-        if not get_child_scores_from_cache:
-            for child in children:
-                self.compute_result_dict(child)
-        candidates = [self[child] for child in children if self[child] is not None]
-        return argmax(candidates, lambda node_value_pair: node_value_pair.expansion_value)
-
-    def compute_result_dict(self, node):
-        if node not in self.tree:
-            self[node] = Expansion(node, self.expansion_value_function(node, self.tree, self.full_tree))
-        else:
-            self[node] = self.__compute_single_node(node, False)
-
-    def update_value(self, target_node, child_value):
-        ancestor_value = self.result_dict.get(id(target_node))
-        assert(ancestor_value is not None)
-        if child_value is None or child_value.expansion_value <= ancestor_value.expansion_value:
-            self[target_node] = self.__compute_single_node(target_node, True)
+        self[node] = self.compute(node)
+
+    def invalidate(self, node):
+        """
+        Invalidate cached value for `node` and its ancestors.
+        """
+        while node is not None:
+            if node in self:
+                del self.result_dict[id(node)]
+            node = self.tree.parent(node)
+
+
+class MaximumLeafValueComputation(TreeComputation):
+
+    def __init__(self, tree: Tree, leaf_value_function, argmax_key=None):
+        super().__init__(tree)
+        self.leaf_value_function = leaf_value_function
+        self.argmax_key = argmax_key
+        self.compute_result_dict(tree.root)
+
+    def compute(self, node):
+        children = self.tree.children(node)
+        if empty(children):
+            return self.leaf_value_function(node, self.tree)
         else:
-            self[target_node] = child_value
-
-    def expand_partial_tree_and_recompute(self, expansion_root):
-        expand_to_node = self[expansion_root].node
-        parent = self.full_tree.parent(expand_to_node)
-        assert(parent is not None)
-        self.tree.add_edge(parent, expand_to_node)
-
-        # Compute available children's expansion results and update value of expanded node
-        # to the highest available child result, then propagate through ancestors:
-        self.recompute_and_propagate_result_to_ancestors(expand_to_node)
-
-    def is_complete(self):
-        return all(self.result_dict[k] is None for k in self.result_dict)
+            candidates = [self[child] for child in children if self[child] is not None]
```

Comment:
What happens if this list turns out to be empty?

Reply:
We rewrote `argmax` to return `None` on an empty collection, so this should be fine.